File: src/services/translationService.ts
class TranslationService {
  private recognition: SpeechRecognition | null = null;

  constructor() {
    if ('SpeechRecognition' in window || 'webkitSpeechRecognition' in window) {
      const SpeechRecognition = window.SpeechRecognition || window.webkitSpeechRecognition;
      this.recognition = new SpeechRecognition();
      this.recognition.continuous = true;
      this.recognition.interimResults = true;
      this.recognition.lang = 'en-US';
    }
  }

  setLanguage(language: string) {
    if (this.recognition) {
      this.recognition.lang = language;
    }
  }

  startRecognition(
    onResult: (text: string) => void,
    onError: (error: Error) => void
  ) {
    if (!this.recognition) {
      onError(new Error('Speech recognition not supported'));
      return;
    }

    this.recognition.onresult = (event) => {
      const transcript = Array.from(event.results)
        .map(result => result[0])
        .map(result => result.transcript)
        .join('');

      onResult(transcript);
    };

    this.recognition.onerror = (event) => {
      onError(new Error(event.error));
    };

    this.recognition.start();
  }

  stopRecognition() {
    if (this.recognition) {
      this.recognition.stop();
    }
  }

  async translate(text: string, targetLanguage: string): Promise<string> {
    try {
      // TODO: Replace with actual translation API call
      // For now, we'll use a mock translation
      const mockTranslations: Record<string, string> = {
        'es': 'Texto traducido al español',
        'fr': 'Texte traduit en français',
        'de': 'Text ins Deutsche übersetzt',
        'zh': '翻译成中文的文本',
        'ar': 'نص مترجم إلى العربية'
      };

      return mockTranslations[targetLanguage] || text;
    } catch (error) {
      console.error('Translation error:', error);
      throw error;
    }
  }
}

export const translationService = new TranslationService(); 

File: backend/services/translation_service.py
import speech_recognition as sr
from typing import Callable, Optional
import uuid

class TranslationService:
    def __init__(self):
        self.recognition = sr.Recognizer()
        self.current_language = 'en-US'
        self.is_listening = False

    def set_language(self, language: str) -> None:
        """Set the language for speech recognition."""
        self.current_language = language

    def start_recognition(
        self,
        on_result: Callable[[str], None],
        on_error: Callable[[Exception], None]
    ) -> None:
        """Start speech recognition."""
        try:
            with sr.Microphone() as source:
                self.is_listening = True
                while self.is_listening:
                    try:
                        audio = self.recognition.listen(source)
                        text = self.recognition.recognize_google(
                            audio,
                            language=self.current_language
                        )
                        on_result(text)
                    except sr.UnknownValueError:
                        continue
                    except sr.RequestError as e:
                        on_error(e)
        except Exception as e:
            on_error(e)

    def stop_recognition(self) -> None:
        """Stop speech recognition."""
        self.is_listening = False

    async def translate(self, text: str, target_language: str) -> str:
        """Translate text to target language."""
        try:
            # TODO: Replace with actual translation API call
            # For now, we'll use a mock translation
            mock_translations = {
                'es': 'Texto traducido al español',
                'fr': 'Texte traduit en français',
                'de': 'Text ins Deutsche übersetzt',
                'zh': '翻译成中文的文本',
                'ar': 'نص مترجم إلى العربية'
            }

            return mock_translations.get(target_language, text)
        except Exception as e:
            print(f'Translation error: {e}')
            raise e

# Create a singleton instance
translation_service = TranslationService() 